### DISCUSSION TITLE
It seems that builder-webpack5 incorrectly builds iframe.html and I have no clue why

### DISCUSSION THREAD

**Discussion by dimafirsov** [storybookjs/storybook#23757] [General] [answered]
Hi all. I'm not posting this as a bug because I don't have a meaningful reproduction, it happens only on my corporate project which, obviously, I can not share. But I can share some configurations and, hopefully, you guys may guide me where and what I did wrong.

So, it all started with the migration to Storybook 7 from version 6, and right from that point webpack5 builder didn't work for me (apparently, it worked before for version 6). I didn't have time to debug and so I used vite builder and it worked. However though with the vite builder the dev ex suffered a bit compared to what we had with webpack, so now I'm trying to get the webpack builder back.

Briefly, what happens is the Storybook build generates `iframe.html` with incorrect contents that looks like a javascript that was intended to generate this page, but was never executed. Looks like this:

<img width="1237" alt="image" src="https://github.com/storybookjs/storybook/assets/16021526/dc12d150-27f7-4b7b-bf83-10ea50fb2c60">

Apparently, storybook starts with a failure because it can't load such weird content.
<img width="2762" alt="image" src="https://github.com/storybookjs/storybook/assets/16021526/89e40475-0adf-4cfa-a6cd-19bb35dc00fe">

Now, I'll post the current config I have for the entire stuff.

# main.ts

```ts
import type { StorybookConfig } from '@storybook/core-webpack';

const path = require('path')

const config: StorybookConfig = {
  features: {
    storyStoreV7: false,
    legacyMdx1: false,
  },
  stories: [
    '../docs/**/*.stories.mdx',
    '../docs/**/*.stories.@(js|jsx|ts|tsx)',
    '../src/**/*.stories.mdx',
    '../src/**/*.stories.@(js|jsx|ts|tsx)',
  ],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/addon-interactions',
    '@storybook/addon-a11y',
    '@geometricpanda/storybook-addon-badges',
    'storybook-addon-designs',
    'storybook-addon-pseudo-states',
    '@storybook/addon-mdx-gfm',
  ],
  staticDirs: [
    {
      from: '../src/assets',
      to: 'assets',
    },
    'static',
  ],
  webpackFinal: async (config) => {
    return {
      ...config,
      resolve: {
        ...config.resolve,
        alias: {
          ...config?.resolve?.alias,
          '@emotion/core': path.join(process.cwd(), 'node_modules/@emotion/react'),
          '@emotion/styled': path.join(process.cwd(), 'node_modules/@emotion/styled'),
          // Ignore the <another_domain>, it's just a cover for some data I can't expose
          '@<another_domain>': path.join(process.cwd(), 'src/index'),
          '@<another_domain>': path.join(process.cwd(), '.storybook/utils.tsx'),
          'emotion-theming': path.join(process.cwd(), 'node_modules/@emotion/react'),
          src: path.join(process.cwd(), 'src/'),
        },
      },
      module: {
        // This section allows to support ?raw syntax
        // more info https://github.com/webpack/webpack/issues/12900#issuecomment-1479392726
        ...config.module,
        rules: [
          ...(config.module as any).rules.map((rule) => ({
            ...rule,
            resourceQuery: { not: [/raw/] },
          })),
          {
            resourceQuery: /raw/,
            type: 'asset/source',
          },
        ],
      },
    }
  },
  framework: {
    name: '@storybook/react-webpack5',
    options: {},
  },
  docs: {
    autodocs: true,
  },
}

export default config
```

# .babelrc.json
```
{
  "sourceType": "unambiguous",
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "chrome": 100
        }
      }
    ],
    "@babel/preset-typescript",
    "@babel/preset-react"
  ],
  "plugins": []
}

```

The `preview.tsx` sets some `parameters` and `decorators`, nothing fancy so I'm not showing it.

Using the latest Storybook, but this has been the case for earlier versions of storybook 7.

What information should I additionally share to provide a better view on the problem?

I tried different versions of storybook 7, and different webpack options, including sending a custom template to it (using `previewMainTemplate`) - nothing worked.

I also saw a template the builder-webpack5 uses (it seems like it does) in `node_modules/@storybook/builder-webpack5/templates/preview.ejs`, and my guess is the builder somehow fails to parse and use it correctly, but I'm not sure. I would like to know more about how the entire thing is working. Would appreciate any guidance.

What could be the problem here, guys? What may cause this file to be generated this way?

@shilman may be you've seen something like this before?

**Comment by valentinpalkovic** [comment]
I am not sure, whether the custom `raw` configuration even needs to be applied, since we are already configuring it [here](https://github.com/storybookjs/storybook/blob/next/code/builders/builder-webpack5/src/preview/base-webpack.config.ts#L81). So the `?raw` syntax should be supported out of the box.

**Reply by ITsvetkoFF** [reply, marked as answer]
@valentinpalkovic that's true. thanks for answer! we've noticed that and were able to tune our webpack config by removing handling of raw assets, but we needed also to tune the rules that are included (otherwise the raw tsx files were included in the docs as transpiled js files and we needed typescript sources)

```javascript
[
...(config?.module?.rules as any).map((rule: any) => {
            return rule.type === 'asset/source'
              ? rule
              : {
                  ...rule,
                  // This is needed to not apply any rules for ?raw files
                  // (otherwise components will be displayed as transpiled)
                  resourceQuery: { not: [/raw/] },
                }
          }),
]
```


One more find that probably deserves a documentation improvement/note is the fact that currently `?raw` syntax won't work in `*.story.ts` files.

**Comment by dimafirsov** [comment]
Thanks for your answer @valentinpalkovic. According to the code you shared you only support `?raw` syntax for `.md` files. We use this syntax for `.story.tsx` files to share their contents in our code examples. We use `.story.tsx` files as holders of a single large story that we then put into `.stories.tsx` files which accumulate all the stories.
